dev/core/test/com/google/gwt/dev/jjs/impl/AnnotationCompilerTest.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2024 GWT Project Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.gwt.dev.jjs.impl;
+
+import com.google.gwt.core.ext.UnableToCompleteException;
+import com.google.gwt.thirdparty.guava.common.base.Joiner;
+
+public class AnnotationCompilerTest extends FullCompileTestBase {
+  @Override
+  protected void optimizeJava() {
+  }
+
+  public void testUseAnnotationWithoutSource() throws UnableToCompleteException {
+    // The JDT represents annotations as three kinds: markers, which have no members,
+    // single-member, which have only one member, and normal, which have more than one.
+    // This test validates that all three work in two cases each: block scope, and class
+    // scope.
+
+    String annotations = Joiner.on(" ").join(
+        "@SampleBytecodeOnlyMarkerAnnotation",
+        "@SampleBytecodeOnlySingleMemberAnnotation(\"abc\")",
+        "@SampleBytecodeOnlyNormalAnnotation(a=1, b=2)"
+    );
+
+    String code = Joiner.on('\n').join(
+        "package test;",
+        "import com.google.gwt.dev.jjs.impl.*;",
+        "public class EntryPoint<T extends " + annotations + " Object> {",
```

Comment:
Should we go all-in here and add annotations everywhere allowed in source code so the test is basically a catch all test and not tailored to the available methods of JDT ASTVisitor (BinaryTypeReferenceVisitor)? 

JDT knows additional scopes (Module, CompilationUnit, Method) and if we update JDT some visitor might change and we might not detect it here given visitors have default implementations. For example block scope could become method scope.

Reply:
Worth considering, but the ASTVisitor doesn't check those, so the particular code in question that is actively looking for binary type refs so that it can complain about them in the logs (and throw an exception) isn't going to fail, at least as I read it.

Would you contribute a test that does that? I'm guessing there is an example out there of all the different ways you can stack annotations that can be used to extend this. We'd need (at least) three instances of every location, to hit marker/normal/single.